src/ipc/Queue.h

```diff
@@ -410,6 +426,96 @@ OneToOneUniQueue::push(const Value &value, QueueReader *const reader)
     return wasEmpty && (!reader || reader->raiseSignal());
 }
 
+template <class Value>
+void
+OneToOneUniQueue::statIn(StoreEntry &entry) const
+{
+    // Nobody can modify our theIn so, after capturing some valid theSize value
+    // in count, we can reliably report all [theIn-count, theIn) items that were
+    // queued at theSize capturing time. We may report items already pop()ed by
+    // the other side, but that is OK because pop() does not modify items -- it
+    // only increments theOut.
+    const auto count = theSize.load();
+    statOpen(entry, "pushIndex", "other", count);
+    statClose<Value>(entry, theIn - count, count); // Branch XXX: theIn-count may get negative!
+}
+
+template <class Value>
+void
+OneToOneUniQueue::statOut(StoreEntry &entry) const
+{
+    // Nobody can modify our theOut so, after capturing some valid theSize value
+    // in count, we can reliably report all [theOut, theOut+count) items that
+    // were queued at theSize capturing time. We will miss new items push()ed by
+    // the other side, but that is OK.
+    const auto count = theSize.load();
+    statOpen(entry, "other", "popIndex", count);
+    statClose<Value>(entry, theOut, count);
+}
+
+/// start cache manager reporting (by reporting queue parameters)
+/// The labels reflect whether the caller owns theIn or theOut data member and,
+/// hence, can report the corresponding value reliably.
+inline void
+OneToOneUniQueue::statOpen(StoreEntry &entry, const char *inLabel, const char *outLabel, const uint32_t count) const
+{
+    // Branch XXX: Some arguments are not %u (i.e. unsigned int)!
+    storeAppendPrintf(&entry, "{ size: %u, capacity: %u, %s: %u, %s: %u",
+                      count, theCapacity, inLabel, theIn, outLabel, theOut);
+}
+
+/// end cache manager reporting with a sample of [start, start + size) items
+template <class Value>
+void
+OneToOneUniQueue::statClose(StoreEntry &entry, const unsigned int start, const uint32_t count) const
+{
+    if (!empty()) {
```

Comment:
I would place the shorter 'else' block first (e.g., see [advice](https://holub.com/goodies/rules.html) 57).

Reply:
To be honest, I do not understand the motivation behind the linked advice. Reading that book (which I do not have) may change my mind, but it feels bogus/made-up at the first glance.

Naturally, there are other contradictory recommendations, which I do understand and find valuable:

* "Put the normal case after the `if` rather than after the `else`"  --Steve McConnell, "Code Complete, 2nd ed.", p356 (according to https://stackoverflow.com/a/166687). I have another edition of that book. AFAICT, the given rationale -- place code close to the corresponding decision -- is bogus, but I speculate that most readers would find if-good-else-bad order more natural than if-bad-else-good.

* "Use the simplest condition for the `if`" -- an application of the KISS and similar rules of thumb

* "Handle exceptional cases first (and return after each) so the rest of the function can safely streamline/focus on processing of the remaining primary case" -- I use this principle a lot, but I am sure others have invented it, and there ought to be better formulations. Many authors, including Steve McConnell, do recommend streamlining the primary processing, of course.

To polish this, I suspect that we need to move the last storeAppendPrintf() outside statClose() so that the exceptional empty() case can be handled first (by printing space and returning), using the simplest condition possible. This refactoring will probably affect function names and might benefit from printing the heading inside the statIn/Out() methods.

It is your call whether to polish this further.

Reply:
Done at 69c619a.